File: api/app/schemas/flashcard.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class FlashcardConfig(BaseModel):
    count: int = Field(ge=5, le=50)
    scope: Literal["whole_video", "topics"]
    segment_ids: list[uuid.UUID] = []
    difficulty: Literal["easy", "medium", "hard", "mixed"] = "mixed"
    style: Literal["definition", "concept", "detail", "mixed"] = "mixed"
    title: str | None = None

    @model_validator(mode="after")
    def _segment_ids_required_for_topics_scope(self) -> "FlashcardConfig":
        if self.scope == "topics" and not self.segment_ids:
            raise ValueError('segment_ids is required when scope is "topics"')
        return self


class FlashcardOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    front: str
    back: str
    segment_id: uuid.UUID | None
    source_start_time: float | None
    difficulty: str | None
    order_index: int


class FlashcardDeckOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    video_id: uuid.UUID
    title: str
    config: dict
    created_at: datetime
    cards: list[FlashcardOut] = []


class FlashcardDeckListItem(BaseModel):
    id: uuid.UUID
    video_id: uuid.UUID
    title: str
    config: dict
    created_at: datetime
    card_count: int

File: web/src/types/flashcard.ts
export type FlashcardScope = 'whole_video' | 'topics'
export type FlashcardDifficulty = 'easy' | 'medium' | 'hard' | 'mixed'
export type FlashcardStyle = 'definition' | 'concept' | 'detail' | 'mixed'

export interface FlashcardConfig {
  count: number
  scope: FlashcardScope
  segment_ids?: string[]
  difficulty?: FlashcardDifficulty
  style?: FlashcardStyle
  title?: string
}

export interface Flashcard {
  id: string
  front: string
  back: string
  segment_id: string | null
  source_start_time: number | null
  difficulty: string | null
  order_index: number
}

export interface FlashcardDeck {
  id: string
  video_id: string
  title: string
  config: Record<string, unknown>
  created_at: string
  cards: Flashcard[]
}

export interface FlashcardDeckListItem {
  id: string
  video_id: string
  title: string
  config: Record<string, unknown>
  created_at: string
  card_count: number
}
